### ISSUE TITLE
name_scope has an effect on names of Variable create by tf.keras.layers.*

### ISSUE THREAD

**Issue by manipopopo** [tensorflow/tensorflow#22142]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.11.0-dev20180907
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

`name_scope` has an effect on names of `Variable`s created by `tf.keras.layers.*`. The following code got different result on tf 1.8 and tf >= 1.9. 

### Source code / logs

```python
import tensorflow as tf

with tf.name_scope('name_scope'):
    tf.keras.layers.Conv2D(3, 3, name='keras')(tf.ones([1, 3, 3, 1]))
    tf.layers.Conv2D(3, 3, name='legacy')(tf.ones([1, 3, 3, 1]))
print([v.name for v in tf.global_variables()])
```

On tf 1.11.0-dev20180907 it prints
```
['name_scope/keras/kernel:0', 
 'name_scope/keras/bias:0',
 'legacy/kernel:0', 
 'legacy/bias:0']
```

On tf 1.8 it prints
```
['keras/kernel:0',
 'keras/bias:0',
 'legacy/kernel:0',
 'legacy/bias:0']
```

**Comment by facaiy** [comment]
@fchollet could you please comment?

**Comment by yanboliang** [comment]
I think this behavior change introduced at https://github.com/tensorflow/tensorflow/commit/693b339ab2f062ec5bbb29f976c5d1fd94fbffa5 . @manipopopo Do you have some issues wrt this change? Thanks.

**Comment by manipopopo** [comment]
@yanboliang 
The following code generates video frames' features in two different ways.

It calls `Conv2D` once to get all the frames' features during training. However, the input videos for evaluation could be very long, it uses `tf.while_loop` to apply `Conv2D` frame-by-frame so that the model can be placed on a GPU with limited memory.

`tf.while_loop` creates a name scope `while` which changes the names of `Conv2D` weights. We can't directly load the model checkpoints for evaluation without specifying `var_list` for `tf.train.Saver`.

```python
import tensorflow as tf


BATCH_SIZE = 1
NUM_FRAMES = 5

CONV_FILTERS = 32


def build_graph_train(video_frames):
  model = tf.keras.layers.Conv2D(CONV_FILTERS, 3, padding='same')
  unfolded_video_frames = tf.reshape(video_frames, [-1, 720, 1080, 3])
  frames_features = tf.reshape(
      model(unfolded_video_frames),
      [BATCH_SIZE, NUM_FRAMES, 720, 1080, CONV_FILTERS])

  # optimize loss(frames_features)
  # ...


def build_graph_evaluate(video_frames):
  model = tf.keras.layers.Conv2D(CONV_FILTERS, 3, padding='same')

  num_frames = tf.shape(video_frames)[1]
  frame_ta = tf.TensorArray(video_frames.dtype, size=num_frames)
  frame_ta = frame_ta.unstack(tf.transpose(video_frames, [1, 0, 2, 3, 4]))

  def process_frame_by_frame(frame_index, frame_features_ta):
    frame = frame_ta.read(frame_index)
    # the shape of frame is [BATCH_SIZE, 720, 1080, 3]
    return frame_index + 1, frame_features_ta.write(frame_index, model(frame))

  _, frame_features_ta = tf.while_loop(
      lambda frame_index, *_: frame_index < num_frames,
      process_frame_by_frame,
      (0, tf.TensorArray(video_frames.dtype, size=num_frames)),
      back_prop=False)
  frame_features = tf.transpose(frame_features_ta.stack(), [1, 0, 2, 3, 4])

  # predict(frame_features)
  # ...


if __name__ == '__main__':
  with tf.Graph().as_default():
    video_frames = tf.zeros([BATCH_SIZE, NUM_FRAMES, 720, 1080, 3])
    build_graph_train(video_frames)

    print(tf.global_variables())
    # [<tf.Variable 'conv2d/kernel:0' shape=(3, 3, 3, 32) dtype=float32>,
    #  <tf.Variable 'conv2d/bias:0' shape=(32,) dtype=float32>]

  with tf.Graph().as_default():
    video_frames = tf.zeros([BATCH_SIZE, NUM_FRAMES, 720, 1080, 3])
    build_graph_evaluate(video_frames)
    print(tf.global_variables())
    # [<tf.Variable 'while/conv2d/kernel:0' shape=(3, 3, 3, 32) dtype=float32>,
    #  <tf.Variable 'while/conv2d/bias:0' shape=(32,) dtype=float32>]
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
